iterator_test.go

```diff
@@ -2737,3 +2739,200 @@ func BenchmarkSeekPrefixTombstones(b *testing.B) {
+	// subbenchmark.
+	var initOps = []int{400_000, 800_000, 2_000_000, 5_000_000, 10_000_000, 50_000_000}
+	// We vary the value size to identify how compaction behaves when the
+	// relative sizes of tombstones and the keys they delete are different.
+	var valueSizes = []int{128, 2048}
+
+	// These should be large enough to assign a unique key to each item in the
+	// queue.
+	const maxQueueIDLen = 1
+	const maxItemLen = 7
+	const maxKeyLen = maxQueueIDLen + 1 + maxItemLen
+	queueIDKeyspace := testkeys.Alpha(maxQueueIDLen)
+	itemKeyspace := testkeys.Alpha(maxItemLen)
+	key := make([]byte, maxKeyLen+1+len(fmt.Sprint(b.N)))
+	val := make([]byte, valueSizes[len(valueSizes)-1])
+
+	getKey := func(q int, i int) []byte {
+		n := testkeys.WriteKey(key, queueIDKeyspace, int64(q))
+		key[n] = '/'
+		prefixLen := n + 1
+		n = testkeys.WriteKey(key[prefixLen:], itemKeyspace, int64(i))
+		return key[:prefixLen+n]
+	}
+
+	type Queue struct {
+		start int
+		end   int // exclusive
+	}
+	var queues = make([]*Queue, queueCount)
+	for i := 0; i < queueCount; i++ {
+		queues[i] = &Queue{}
+	}
+
+	processQueueOnce := func(d *DB, batch *Batch, rng *rand.Rand, deleteRatio float32, valueSize int, batchSize int) *Batch {
+		for {
+			// Randomly pick a queue to process.
+			q := rng.Intn(queueCount)
+			queue := queues[q]
+
+			isDelete := rng.Float32() < deleteRatio
+
+			if isDelete {
+				// Only process the queue if it's not empty. Otherwise, retry
+				// with a different queue.
+				if queue.start < queue.end {
+					require.NoError(b, batch.Delete(getKey(q, queue.start), nil))
+					queue.start++
+					break
+				}
+			} else {
+				// Append to the queue.
+				require.NoError(b, batch.Set(getKey(q, queue.end), val[:valueSize], nil))
+				queue.end++
+				break
+			}
+		}
+
+		if batch.Len() >= batchSize {
+			require.NoError(b, batch.Commit(NoSync))
+			d.AsyncFlush()
+			return d.NewBatch()
+		}
+		return batch
+	}
+
+	type RunOptions struct {
+		deleteRatio float32
+		initialOps  int
+		valueSize   int
+	}
+
+	// Initialize a new DB instance for each {deleteRatio, numInitOps,
+	// valueSize} configuration.
+	dbs := make([]*DB, 0)
+	runOpts := make([]RunOptions, 0)
+	var grp errgroup.Group
+	var numInitialized atomic.Uint32
+	for _, deleteRatio := range deleteRatios {
+		for _, valueSize := range valueSizes {
+			for _, numInitOps := range initOps {
+				o := (&Options{
+					DisableWAL:         true,
+					FS:                 vfs.NewMem(),
+					Comparer:           testkeys.Comparer,
+					FormatMajorVersion: FormatNewest,
+				}).EnsureDefaults()
+				opts := RunOptions{
+					deleteRatio: deleteRatio,
+					initialOps:  numInitOps,
+					valueSize:   valueSize,
+				}
+				runOpts = append(runOpts, opts)
+
+				d, err := Open("", o)
+				require.NoError(b, err)
+				dbs = append(dbs, d)
+
+				// First we process queues initialOps times (parallelized to
+				// speed things up).
+				grp.Go(func() error {
```

Comment:
You're speeding up the initialization for all the different sub-benchmarks, which you wouldn't get if you moved it all into a helper function called once, but it's helpful to have each benchmark get as much in the way of system resources (eg. memory) as it could. It's also possible that you aren't interested in running some of the sub-benchmarks (eg. by passing a regex filter to `go test -bench`, which applies to the benchmark names) but want to run or re-run others. Currently you'll initialize them all even if you only use some of these in practice further down. So while it's a bit of a tradeoff to serialize initialization, it's a worthwhile one IMO.

Reply:
For future testing, I'll probably focus on the `deleteRatio=0.1, valueSize=2048` config because it seems to be a clear example of tombstone buildup.